CyberArk InitialSupport

cyberark-bizdev opened the pull request:
Added cyberarkpassword lookup plugin: It allows to retrieve credentials
(password, sshkey) from CyberArk Digital Vault

Added Modules:
- cyberark_authentication: Logon/Logoff to CyberArk Vault
- cyberark_user: user management

These 2 modules use CyberArk Privileged Account Security Web Services
SDK

cyberark-bizdev committed:
Added cyberarkpassword lookup plugin

Added cyberarkpassword lookup plugin: It allows to retrieve credentials
(password, sshkey) from CyberArk Digital Vault

cyberark-bizdev committed:
Added Modules: cyberark_authentication & cyberark_user

Added Modules:
- cyberark_authentication: Logon/Logoff to CyberArk Vault
- cyberark_user: user management

These 2 modules use CyberArk Privileged Account Security Web Services
SDK

cyberark-bizdev committed:
Update cyberark_authentication.py

cyberark-bizdev committed:
Update cyberark_user.py

enunez-cyberark commented:
!needs_rebase

enunez-cyberark commented:
!needs_rebase

enunez-cyberark commented:
!needs_rebase

cyberark-bizdev committed:
Removed ternary conditional to comply with 2.4

cyberark-bizdev committed:
Replaced usage of iteritems() for items() to comply with python3

cyberark-bizdev committed:
PEP8 Updates

cyberark-bizdev committed:
Fixed Doc Issues

cyberark-bizdev committed:
Doc Fixes

cyberark-bizdev committed:
More Doc Fixes

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+            sample: true
+'''
+
+from ansible.module_utils.basic import *
```
Please only import what you need, rather than `*`

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+        "validateCerts": {"required": False, "type": "bool",
+                          "default": "false"},
+        "username": {"required": False, "type": "str"},
+        "password": {"required": False, "type": "str"},
```
no_log needs to be set to True for all password fields.

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+                          "default": "false"},
+        "username": {"required": False, "type": "str"},
+        "password": {"required": False, "type": "str"},
+        "newPassword": {"required": False, "type": "str"},
```
no_log needs to be set to True for all password fields.

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+            - A string containing the base URL of the server hosting CyberArk's Privileged Account Security Web Services SDK
+    validateCerts:
+        required: False
+        default: false
```
Default should be True (needs updating in code as well).

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+        description:
+            - Specifies if an authentication logon/logoff and a cyberarkSession should be added/removed
+    userName:
+        required: False
```
All `required: False` lines can be removed.

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+        required: False
+        description:
+            - The password of the user
+    newPassword:
```
Can you still create idempotent (playbooks that on the second run don't report any changes) using `password` and `newPassword`?

cyberark-bizdev replied:
passing newPassword is a way to request a password change ... I added logic to everytime newPassword is sent to set changed to True... with newPassword it will report changes

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+        required: False
+        description:
+            - A string containing the base URL of the server hosting CyberArk's Privileged Account Security Web Services SDK
+    validateCerts:
```
`validate_certs:` to be consistent with other modules

cyberark-bizdev replied:
Done

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+            - Dictionary set by a CyberArk authentication containing the different values to perform actions on a logged-on CyberArk session
+'''
+
+EXAMPLES = '''
```
For these examples to be more "real world" do they need to `register` the returned token and use that with another module?

cyberark-bizdev replied:
in the case of the module cyberark_authentication... cyberarkSession is returned as a fact ... and you can use it as it in the example for the logoff where cyberarkSession is "{{ cyberarkSession }}"

gundalow reviewed (changes requested):
I've only looked at the first module.

As per http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html each PR must only contain a single module.

Please remove the other modules from this PR and wait for the first one to be reviewed and merged before raising other PRs for the rest.

gundalow commented on lib/ansible/modules/privileged_account_security/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,292 @@
+
+def main():
+
+    fields = {
```
Will need updating based on DOCUMENTATION updates

cyberark-bizdev replied:
Done

thaumos commented:
![](https://github.trello.services/images/mini-trello-icon.png) [CyberArk - Lookup Plugin](https://trello.com/c/P3WRtYhE/186-cyberark-lookup-plugin)

gundalow commented:
I've only looked at the first module.

As per http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html each PR must only contain a single module.

Please remove the other modules from this PR and wait for the first one to be reviewed and merged before raising other PRs for the rest.

cyberark-bizdev committed:
Removing cyberark_user module, and cyberark lookup plugin

for initial approval of PR, and continue with 2 different PRs after PR
#21764 is approved.

cyberark-bizdev committed:
PEP8 Fixes

cyberark-bizdev reviewed (commented):
Updated based on feedback. I made the PR to be about 1 single file at the moment, I will submit separate PRs after this one gets approved.

enunez-cyberark commented:
Done… I made changes based on feedback, also submitted only one module for PR #21764.



-------
Edward Nunez | CyberArk | Business Development | Alliances Technology Expert | +1(510)909-7329




On Feb 22, 2017, at 9:59 AM, John R Barker <notifications@github.com<mailto:notifications@github.com>> wrote:


I've only looked at the first module.

As per http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html<https://urldefense.proofpoint.com/v2/url?u=http-3A__docs.ansible.com_ansible_dev-5Fguide_developing-5Fmodules-5Fin-5Fgroups.html&d=DwMFaQ&c=c22L3idIg_s6WyfhOrpwxHcMGUqKswlQ1WoVlCpWXQs&r=_dPs6ec6_84caeFi7hBwUG15E6rfKbMObhkaANGFCwg&m=npBq1aC60w3P0-hvRvxXNL1vwgk3YZAL8GNNVasuJsU&s=WxgJchFVSVfHpN0awwCWS8JFT_nwhJGOkiU6os6YR9c&e=> each PR must only contain a single module.

Please remove the other modules from this PR and wait for the first one to be reviewed and merged before raising other PRs for the rest.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ansible_ansible_pull_21764-23issuecomment-2D281749364&d=DwMFaQ&c=c22L3idIg_s6WyfhOrpwxHcMGUqKswlQ1WoVlCpWXQs&r=_dPs6ec6_84caeFi7hBwUG15E6rfKbMObhkaANGFCwg&m=npBq1aC60w3P0-hvRvxXNL1vwgk3YZAL8GNNVasuJsU&s=P9sxBfMq1R1WoRWOU8DYwWlKgondOwf3QA7FGxacZ2c&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AYQpP9XTLV2FzLgLlJWHkgMsZJYDW4s-2Dks5rfHd5gaJpZM4MIQR0&d=DwMFaQ&c=c22L3idIg_s6WyfhOrpwxHcMGUqKswlQ1WoVlCpWXQs&r=_dPs6ec6_84caeFi7hBwUG15E6rfKbMObhkaANGFCwg&m=npBq1aC60w3P0-hvRvxXNL1vwgk3YZAL8GNNVasuJsU&s=dfWTuZIc3GZN_bsXaoxqhCqTNjKczFlER-GByX1PV4E&e=>.

_______________________________________________
This e-mail may contain information that is confidential, privileged or otherwise protected from disclosure.
If you are not an intended recipient of this e-mail, do not duplicate or redistribute it by any means. Please delete it and any attachments and notify the sender that you have received it in error.

enunez-cyberark commented:
Please go ahead and review the updates

enunez-cyberark commented:
shipit

erasmix commented:
I downloaded it and tested it on my environment. It works OK.

erasmix commented:
shipit

cyberark-bizdev committed:
Moved cyberark modules to identity category

From IRC #ansible-devel recommendation (@bcoca) I moved cyberark to
identity category so the authorized maintainers can provide feedback
and move it forward.

enunez-cyberark commented:
!needs_rebase

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+        default: present
+        choices: ['present', 'absent']
+        description:
+            - Specifies if an authentication logon/logoff and a cyberarkSession should be added/removed
```
All description lines (apart from short_description) should have full stops.
Must be updated in all description fields.
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+        choices: ['present', 'absent']
+        description:
+            - Specifies if an authentication logon/logoff and a cyberarkSession should be added/removed
+    userName:
```
Choose between camel case (here for userName) and delimiter-separated words (in validate_certs)
Most of ansible is based on delimiter-separated words.

cyberark-bizdev replied:
I was initially camel case, but on a previous review validateCerts was changed to validate_certs to be consistent in the call to open_url.

I will change the whole file to be delimiter-separated.

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+description:
+    - "Authenticates to CyberArk Vault using Privileged Account Security Web Services SDK and
+       creates a session fact that can be used by other modules. It returns an Ansible fact
+       called I(cyberarkSession). Every module can use this fact as C(cyberarkSession) parameter."
```
Not sure if it's the good way to do this.
I think you better put common code (like authentication) in some module_utils file and use it in modules.
Look at cloudstak modules and module_utils files.

cyberark-bizdev replied:
I took a look at cloudstack modules and module_utils, and I see the need for it based on using a python library and having that as a utility module for it. 

In this case, the functionality of the cyberark modules is based on REST apis available accross the board. 
I based it the same way it was done for module ovirt_auth under modules/cloud/ovirt

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+    newPassword = module.params["newPassword"]
+    useSharedLogonAuthentication = module.params[
+        "useSharedLogonAuthentication"]
+    useRadiusAuthentication = module.params["useRadiusAuthentication"]
```
Camelcase or not ? :)

cyberark-bizdev replied:
:-) ... I am all for either one... initially was camelCase, and then started diverging when I had to change it for validateCerts (validate_certs). 

Changing everything to be consistent.

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+        if (not useSharedLogonAuthentication and (
+                username is None or password is None)):
+            module.fail_json(
+                msg="with state=%s and useSharedLogonAuthentication=%s, " +
```
Prefer use require_one_of in module_spec

cyberark-bizdev replied:
Updated AnsibleModule module_spec parameters to use mutually_exclusive, required_if and required_together parameters to avoid manual validation of the parameters.

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+            payload = json.dumps({"username": module.params.get(
+                "username"), "password": module.params.get("password")})
+
+    else:
```
Could you add some comment for better code readability, your module is hard to read

cyberark-bizdev replied:
Done

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+def main():
+
+    fields = {
+        "apiBaseUrl": {"required": False, "type": "str"},
```
remove required: false 
It's implicit

cyberark-bizdev replied:
Done

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+        "password": {"required": False, "type": "str", "no_log": True},
+        "newPassword": {"required": False, "type": "str", "no_log": True},
+        "useSharedLogonAuthentication": {"default": False, "type": "bool"},
+        "useRadiusAuthentication": {"default": False, "type": "bool"},
```
It may be better to define only one required authentication param like authentication_method with choice between shared and radius
Then add a line in required_if for username and password in case of shared auth.

cyberark-bizdev replied:
In reality the 2 parameters even though are named similar in CyberArk REST api accomplish 2 different things. 
use_radius_authentication is used when passing username and password during logon action (state=present) and not using shared logon (which does not accept username and password). 

I updated the AnsibleModule module_spec parameters to enhance the check of the parameters using mutually_exclusive, required_if and required_together

bjolivot commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,304 @@
+        supports_check_mode=True)
+
+    if module.check_mode:
+        module.exit_json(change=False)
```
Could you test authentication without changing anything ?
It would be helpfull

cyberark-bizdev replied:
@bjolivot thanks for the suggestion, I changed the logic internally that when running in check_mode it can be used to test authentication, and not actually performing password changes when passing a new_password.

cyberark-bizdev committed:
Updates based on community_review by bjolivot

- Updated description lines to have full stops in the documentation
section.
- changed file to use delimiter-separated words instead of camel case
- Updated AnsibleModule module_spec parameters to use
mutually_exclusive, required_if and required_together parameters to
avoid manual validation of the parameters.
- Added comments for more readability.
- Removed “required”: false as they are implicit.
- Enhanced check_mode handling.

cyberark-bizdev committed:
PEP8 Updates

enunez-cyberark commented:
@bjolivot I just pushed commit to address all your review comments. Please let me know if you still see anything not clear.

jarcher commented:
Does this need Vault or AIM as well in order to test?

enunez-cyberark commented:
@jarcher It needs CyberArk Vault and PVWA (Password Vault Web Access) with Web Services SDK. 
- CyberArk's Privileged Account Security WebServices SDK

erasmix commented:
It requires both. Uses the CCP.

From: John Archer [mailto:notifications@github.com]
Sent: Friday, March 3, 2017 9:52 AM
To: ansible/ansible <ansible@noreply.github.com>
Cc: Erasmo Acosta <Erasmo.Acosta@CyberArk.com>; Comment <comment@noreply.github.com>
Subject: Re: [ansible/ansible] CyberArk InitialSupport (#21764)


Does this need Vault or AIM as well in order to test?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ansible_ansible_pull_21764-23issuecomment-2D284022970&d=DwMCaQ&c=c22L3idIg_s6WyfhOrpwxHcMGUqKswlQ1WoVlCpWXQs&r=g83kplevHRIT2H4eOffAXLRKAhQIzzanV2BCDvVLRP8&m=bereaRB7qRRmtwZ30m4UEoUFKjxQxhUVPA2KSYKzq2s&s=YOafbqG6_xB7mMn9tPasQwql10DQ9EkQhI3NGUNdxds&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AYyXO4GXdHZs0nHPz8eMHGJpTki0zjaiks5riFNPgaJpZM4MIQR0&d=DwMCaQ&c=c22L3idIg_s6WyfhOrpwxHcMGUqKswlQ1WoVlCpWXQs&r=g83kplevHRIT2H4eOffAXLRKAhQIzzanV2BCDvVLRP8&m=bereaRB7qRRmtwZ30m4UEoUFKjxQxhUVPA2KSYKzq2s&s=50cdHDoa5Qf9I-h4PBiiowW6Axu-Vnn7JAns-ukaX74&e=>.

_______________________________________________
This e-mail may contain information that is confidential, privileged or otherwise protected from disclosure.
If you are not an intended recipient of this e-mail, do not duplicate or redistribute it by any means. Please delete it and any attachments and notify the sender that you have received it in error.

enunez-cyberark commented:
@erasmix It does NOT use CCP (Central Credential Provider)... but it uses CyberArk's Privileged Account Security WebServices SDK.

cyberark-bizdev commented:
##### SUMMARY
Added Module:

cyberark_authentication: Logon/Logoff to CyberArk Vault

This module uses CyberArk Privileged Account Security Web Services
SDK

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/identity/cyberark/cyberark_authentication.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```


##### ADDITIONAL INFORMATION

jctanner commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+        else:
+            end_point = "/PasswordVault/WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logoff"
+
+    try:
```
having such a large try/except can lead to troubleshooting problems for the end user. Please break it up into smaller blocks.

cyberark-bizdev replied:
Changed to smaller block for try/except

jctanner commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+module: cyberark_authentication
+short_description: "Module for CyberArk Vault Authentication using Privileged Account Security Web Services SDK"
+author: "Edward Nunez (@enunez-cyberark)"
+version_added: "2.3"
```
This should be 2.4 now

cyberark-bizdev replied:
Fixed

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+            module.fail_json(
+                msg="error in end_point=>" +
+                end_point +
+                json.dumps(headers))
```
Could/should this return an extra item for headers?

``` python 
     module.fail_json(msg='error in end_point=>' + endpoint, headers=headers)
```

abadger replied:
+1 and the same for all the other fail_json's as well.

cyberark-bizdev replied:
Changed

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+            "] payload=[" +
+            payload +
+            "] ===>" +
+            str(e),
```
split headers and/or payload to different items in the fail_json? 

``` python
    fail_json(msg='asfad', headers=headers, payload=payload)
```

Suppose that depends if it would ever be useful on its own instead of part of the error msg.

cyberark-bizdev replied:
Changed

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+            exception=traceback.format_exc(),
+            status_code=e.code)
+
+    except Exception:
```
I would suggest 'except Exception as e' unless there is a reason for the sys.exc_info()

cyberark-bizdev replied:
Done

abadger commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'committer',
+                    'version': '1.0'}
```
Should be metadata_version

cyberark-bizdev replied:
Fixed

thaumos commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,324 @@
+
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'committer',
```
committer should be change to community

cyberark-bizdev replied:
Changed

gundalow commented:
CI retriggered

enunez-cyberark committed:
Updates based on IRC Feedback June 6

cyberark-bizdev commented:
!needs_info
!needs_template

enunez-cyberark committed:
Fixed description for token item

enunez-cyberark committed:
Fixed Documentation RETURN string

enunez-cyberark committed:
Fixed PEP8 W291 trailing whitespace

sivel commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+        module.fail_json(
+            msg="end_point=" +
+            api_base_url +
+            end_point,
```
Confusing string concatenation.  Look into `printf`-like string formatting or `.format()`

I see other places where this should be changed.

cyberark-bizdev replied:
Fixed.

sivel commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+            if use_shared_logon_authentication:
+                token = json.loads(response.read())["LogonResult"]
+            else:
+                token = json.loads(response.read())["CyberArkLogonResult"]
```
Is there the potential that the response could be 200, but the JSON not include these keys?  Maybe they should be guarded with try/except to avoid exceptions.

alikins replied:
Agreed on exception handling here. response.read() can raise exceptions as well as json.loads() and the item lookup.

cyberark-bizdev replied:
Enclosed block in try/except

sivel commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+                    "api_base_url": api_base_url,
+                    "validate_certs": validate_certs,
+                    "use_shared_logon_authentication":
+                    use_shared_logon_authentication},
```
Our code formatting standards allow for 160 character width, move this onto the same line to make it more readable.

cyberark-bizdev replied:
Recommendation was incorporated.

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+import urllib
+import httplib
+import traceback
+import sys
```
urllib, traceback, sys dont seem to be used.

cyberark-bizdev replied:
remove unused imports

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+            end_point,
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
```
to_text() isnt imported

cyberark-bizdev replied:
import added

alikins commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,301 @@
+        mutually_exclusive=mutually_exclusive,
+        required_if=required_if,
+        required_together=required_together,
+        supports_check_mode=True)
```
I like the code organization here 👍 , but it is a little different than most modules. Some folks may prefer the deeply nested args style. I don't think you should change it, just a heads up that it might be a little unusual to some folks.

cyberark-bizdev replied:
Thanks... I'll keep it as is.

enunez-cyberark committed:
Changes based on feedback from community review

enunez-cyberark committed:
Added import to_text from ansible.module_utils._text

bcoca commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,300 @@
+            msg="end_point=%s%s" % (api_base_url, end_point),
+            payload=payload,
+            headers=headers,
+            exception=to_text(e),
```
A couple of things, we try to avoid 'catchalls' as they don't give good error messages. We try to capture the errors after a specific action "friendly' part that suggests what was being attempted and possibly what user can do to change error to success.

Also, the 'exception' field is badly named for python, it is mostly for tracebacks and we normally append the exception string to the msg field after the 'friendly msg'

cyberark-bizdev replied:
@bcoca I made changes based on your suggestion. I made a friendlier error message, and appended the exception to it.  Renamed also the exception field.

bcoca commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,300 @@
+    returned: success
+    type: dict
+    sample:
+        {
```
if you know the contents of the dict, you can also document them as 'subkeys of the cyberark_session key

cyberark-bizdev replied:
@bcoca I changed the documentation to describe the individual subkeys of the cyberark_session key, instead of the example data I had before.

enunez-cyberark committed:
Updates based on recommendation from community

cyberark-bizdev commented:
@bcoca What can we do to move this to approved so I can submit a 2nd module that depends on this one ?

tima commented:
You need to work with the ansibot to mark this PR  read_for_review. See 
 https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md. Pinging an individual contributor typically slows resolution.

tima commented:
ready_for_review

cyberark-bizdev commented:
@tima thanks ... it was not clear to me the actions I need to trigger ... really appreciate your help.

tima commented:
@cyberark-bizdev: You need to post a comment with just "ready_for_review" in it here. I forgot that only to submitter can trigger the bot to take action with that command. When the bot runs it will label this PR that the core committers will see it.

cyberark-bizdev commented:
ready_for_review

bcoca commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+    except (HTTPError, httplib.HTTPException) as http_exception:
+
+        module.fail_json(
+            msg=("Exception while performing authentication."
```
use Errror instead of exception, most users know what an error is, only programmers understand what exception means in this context

cyberark-bizdev replied:
@bcoca thanks for the feedback. I changed it to show Error instead of exception in all the friendly messages.

enunez-cyberark committed:
Changed Exception for Error in friendly messages in try/except blocks

cyberark-bizdev commented:
ready_for_review

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+              set to C(false) used on personally controlled sites using self-signed
+              certificates.
+    use_shared_logon_authentication:
+        default: false
```
Please use:
```
    type: bool
    default: 'no'
```

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+        description:
+            - Whether or not Shared Logon Authentication will be used.
+    use_radius_authentication:
+        default: false
```
Please use:
```
    type: bool
    default: 'no'
```

cyberark-bizdev replied:
Fixed

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+    api_base_url: "{{ web_services_base_url }}"
+    username: "{{ password_object.password }}"
+    password: "{{ password_object.passprops.username }}"
+    use_shared_logon_authentication: false
```
We prefer to use `yes` and `no` over `true` and `false` for booleans in YAML.

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+DOCUMENTATION = '''
+---
+module: cyberark_authentication
+short_description: "Module for CyberArk Vault Authentication using Privileged Account Security Web Services SDK"
```
You don't need quotes here. We tend to only add quotes to YAML when it is required.
Also make this shorter, you don't have to mention the SDK used in the short_description, leave that for the normal description and/or the notes.

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+---
+module: cyberark_authentication
+short_description: "Module for CyberArk Vault Authentication using Privileged Account Security Web Services SDK"
+author: "Edward Nunez (@enunez-cyberark)"
```
You don't need quotes here. We tend to only add quotes to YAML when it is required.
Also we prefer this to be a list (as there could be more than one author).
So:
```yaml
author:
- Edward Nunez (@enunez-cyberark)
```

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+author: "Edward Nunez (@enunez-cyberark)"
+version_added: "2.4"
+description:
+    - "Authenticates to CyberArk Vault using Privileged Account Security Web Services SDK and
```
Same here, no quotes needed for a multiline string.

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+options:
+    state:
+        default: present
+        choices: ['present', 'absent']
```
No quotes needed for present or absent, so:
```yaml
    choices: [ absent, present ]
```

cyberark-bizdev replied:
Fixed.

dagwieers commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+        description:
+            - A string containing the base URL of the server hosting CyberArk's Privileged Account Security Web Services SDK.
+    validate_certs:
+        default: true
```
Please use:
```
    type: bool
    default: 'yes'
```

cyberark-bizdev replied:
Fixed.

bcoca commented on lib/ansible/modules/identity/cyberark/cyberark_authentication.py:
```diff
@@ -0,0 +1,315 @@
+                 "\n*** end_point=%s%s\n ==> %s" % (api_base_url, end_point, to_text(http_exception))),
+            payload=payload,
+            headers=headers,
+            exception=to_text(http_exception),
```
this field is for tracebacks (i know, badly named for python) since you are already passing the exception message this is both redundant and not the purpose for the field.

same issue in all subsequent fail_json

cyberark-bizdev replied:
Fixed.

thaumos commented:
@cyberark-bizdev and @erasmix any updates to the latest review?

thaumos commented:
ping @cyberark-bizdev and @erasmix anything?

cyberark-bizdev committed:
Updates based on community review (bcoca & dagwieers)

cyberark-bizdev committed:
Fixed httplib for python3 (http.client)

cyberark-bizdev commented:
ready_for_review

cyberark-bizdev commented:
!needs_template
